### PR TITLE
Improve first iteration logic

### PR DESCRIPTION
**Pull request by ikonst**
Improving the first iteration logic in `PageIterator` and `ResultIterator`.

Both iterators had a `_first_iteration` state attribute that could be replaced with better logic:
- In `PageIterator`, it can be replaced with `_is_last_page` which
  - is exception-safe (only mutate after we do I/O), and
  - (IMO) more directly relates to the state we're handling
- In `ResultsIterator`, we can do without `_first_iteration` (but we should initialize `_index` and `_count`)

Inspired by #1059.

### CODE REVIEW

File: pynamodb/pagination.py [hunk 1]
```diff
@@ -170,15 +169,17 @@ def __init__(
         settings: OperationSettings = OperationSettings.default,
     ) -> None:
         self.page_iter: PageIterator = PageIterator(operation, args, kwargs, rate_limit, settings)
-        self._first_iteration = True
         self._map_fn = map_fn
         self._limit = limit
         self._total_count = 0
+        self._items: List = []
```
Comment (anandswaminathan) [inline pynamodb/pagination.py:175]: @ikonst Why is this required. I don't see any calls to `self._items` ahead of https://github.com/pynamodb/PynamoDB/pull/1101/files#diff-05a16c49172d4c74ad60089cd28c1ba68f5da9699bbb94ce1ccee0f9e60d9dfdR182

Comment (ikonst) [reply pynamodb/pagination.py:175]: You are right. And for all practical reasons, `self._items[X]` will raise just the same (for an empty list).

File: pynamodb/pagination.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import time
-from typing import Any, Callable, Dict, Iterable, Iterator, TypeVar, Optional
+from typing import Any, Callable, Dict, Iterable, Iterator, List, Optional, TypeVar
```
Comment (anandswaminathan) [inline pynamodb/pagination.py:2]: Remove "List"

**Review by anandswaminathan — APPROVED** [review]
One small comment

File: pynamodb/pagination.py (hunk 2 above)
Comment (ikonst) [reply pynamodb/pagination.py]: Thanks. We really ought to upgrade to repo to common tooling (black, reorder-python-imports, pyupgrade) but I'm dreading the diff size 😅
